internal/shim-kvm/src/hostcall.rs

```diff
@@ -439,6 +441,80 @@ impl Handler for HostCall<'_> {
         Ok(())
     }
 
+    fn mremap(
+        &mut self,
+        platform: &impl Platform,
+        old_addr: NonNull<c_void>,
+        old_size: c_size_t,
+        new_size: c_size_t,
+        flags: Option<MremapFlags>,
+    ) -> sallyport::Result<NonNull<c_void>> {
+        eprintln!(
+            "SC> mremap({:?}, {:?}, {:?}, {:?}) = …",
+            old_addr, old_size, new_size, flags
+        );
+
+        // Make sure old address range is owned by process
+        let source_slice = platform.validate_slice::<u8>(old_addr.as_ptr() as _, old_size)?;
+
+        let mut target_addr = source_slice.as_ptr();
+
+        target_addr = unsafe {
+            match flags {
+                Some(MremapFlags {
+                    FIXED: None,
+                    DONTUNMAP: false,
+                }) => {
+                    match old_size.cmp(&new_size) {
+                        Ordering::Equal => target_addr,
+                        Ordering::Greater => {
+                            // simply unmap the tail
+                            let addr = target_addr.add(new_size);
+
+                            // It is not an error if the indicated range does not contain any mapped pages.
+                            let _ = self.munmap(
+                                platform,
+                                NonNull::new(addr as *mut c_void).ok_or(EINVAL)?,
+                                old_size.checked_sub(new_size).ok_or(EINVAL)?,
+                            );
+
+                            eprintln!("SC> mremap() shrink = {:?}", target_addr);
+
+                            target_addr
+                        }
+                        Ordering::Less => {
+                            // simply copy the old data to a new location
+
+                            // FIXME: find out the permissions of the old segment
```

Comment:
`mmledger` could be used for this. it does not have query function for this but nothing prevents to add one.